deepchem/models/torch_models/kfac_optimizer.py

```diff
@@ -0,0 +1,477 @@
+    super(KFACOptimizer, self).__init__(model.parameters(), defaults)
+    self.batch_averaged = batch_averaged
+
+    self.known_modules = {'Linear', 'Conv2d'}
+
+    self.modules: List[torch.nn.Module] = []
+
+    self.model = model
+    self._prepare_model()
+
+    self.steps = 0
+
+    self.m_aa: Dict[torch.nn.Module, torch.Tensor] = {}
+    self.m_gg: Dict[torch.nn.Module, torch.Tensor] = {}
+    self.Q_a: Dict[torch.nn.Module, torch.Tensor] = {}
+    self.Q_g: Dict[torch.nn.Module, torch.Tensor] = {}
+    self.d_a: Dict[torch.nn.Module, torch.Tensor] = {}
+    self.d_g: Dict[torch.nn.Module, torch.Tensor] = {}
+    self.stat_decay = stat_decay
+
+    self.kl_clip = kl_clip
+    self.TCov = TCov
+    self.TInv = TInv
+
+    self.mean = mean
+
+  def try_contiguous(self, x: torch.Tensor) -> torch.Tensor:
+    """
+    Checks the memory layout of the input tensor and changes it to contiguous type.
+
+    Parameters:
+    -----------
+    x: torch.Tensor
+      The input tensor to be made contiguous in memory, if it is not so.
+
+    Return:
+    -------
+    torch.Tensor
+      Tensor with contiguous memory
+    """
+    if not x.is_contiguous():
+      x = x.contiguous()
+
+    return x
+
+  def _extract_patches(
+      self, x: torch.Tensor, kernel_size: Tuple[int, ...],
+      stride: Tuple[int, ...], padding: Union[int, str,
+                                              Tuple[int, ...]]) -> torch.Tensor:
+    """
+    Extract patches of a given size from the input tensor given. Used in calculating
+    the matrices for the kronecker product in the case of 2d Convolutions.
+
+    Parameters:
+    -----------
+    x: torch.Tensor
+      The input feature maps. with the size of (batch_size, in_c, h, w)
+    kernel_size: Tuple[int, ...]
+      the kernel size of the conv filter.
+    stride: Tuple[int, ...]
+      the stride of conv operation.
+    padding: Union[int, str, Tuple[int, ...]]
+      number of paddings. be a tuple of two elements
+
+    Return:
+    -------
+    torch.Tensor:
+      Extracted patches with shape (batch_size, out_h, out_w, in_c*kh*kw)
+    """
+    if isinstance(padding, tuple):
+      if padding[0] + padding[1] > 0:
+        x = torch.nn.functional.pad(x, (padding[1], padding[1], padding[0],
+                                        padding[0])).data  # Actually check dims
+    elif isinstance(padding, int):
+      if padding > 0:
+        x = torch.nn.functional.pad(x,
+                                    (padding, padding, padding, padding)).data
+    elif isinstance(padding, str):
+      if padding == 'VALID':
+        pad = int((kernel_size[0] - 1) / 2)
+        x = torch.nn.functional.pad(x, (pad, pad, pad, pad)).data
+
+      x = x.unfold(2, kernel_size[0], stride[0])
+      x = x.unfold(3, kernel_size[1], stride[1])
+      x = x.transpose_(1, 2).transpose_(2, 3).contiguous()
+      x = x.view(x.size(0), x.size(1), x.size(2),
+                 x.size(3) * x.size(4) * x.size(5))
```

Comment:
Are you sure the indentation here is correct?  These lines are only supposed to be executed when `padding` is a string?